src/hdx/utilities/azure_utils.py

```diff
@@ -0,0 +1,197 @@
+            "verb": "GET",
+            "Content-Encoding": "",
+            "Content-Language": "",
+            "Content-Length": "",
+            "Content-MD5": "",
+            "Content-Type": "",
+            "Date": "",
+            "If-Modified-Since": "",
+            "If-Match": "",
+            "If-None-Match": "",
+            "If-Unmodified-Since": "",
+            "Range": "",
+            "CanonicalizedHeaders": "x-ms-date:"
+            + request_time
+            + "\nx-ms-version:"
+            + api_version
+            + "\n",
+            "CanonicalizedResource": "/"
+            + account
+            + "/"
+            + container
+            + "/"
+            + blob,
+        }
+
+        signature = (
+            parameters["verb"]
+            + "\n"
+            + parameters["Content-Encoding"]
+            + "\n"
+            + parameters["Content-Language"]
+            + "\n"
+            + parameters["Content-Length"]
+            + "\n"
+            + parameters["Content-MD5"]
+            + "\n"
+            + parameters["Content-Type"]
+            + "\n"
+            + parameters["Date"]
+            + "\n"
+            + parameters["If-Modified-Since"]
+            + "\n"
+            + parameters["If-Match"]
+            + "\n"
+            + parameters["If-None-Match"]
+            + "\n"
+            + parameters["If-Unmodified-Since"]
+            + "\n"
+            + parameters["Range"]
+            + "\n"
+            + parameters["CanonicalizedHeaders"]
+            + parameters["CanonicalizedResource"]
+        )
+        signed_string = base64.b64encode(
+            hmac.new(
+                base64.b64decode(key),
+                msg=signature.encode("utf-8"),
+                digestmod=hashlib.sha256,
+            ).digest()
+        ).decode()
+
+        headers = {
+            "x-ms-date": request_time,
+            "x-ms-version": api_version,
+            "Authorization": ("SharedKey " + account + ":" + signed_string,),
+        }
+
+        url = (
+            "https://"
+            + account
+            + ".blob.core.windows.net/"
+            + container
+            + "/"
+            + blob
+        )
+
+        if keep and exists(url):
+            print(f"The blob URL exists: {url}")
+            return path
+        self.setup(
+            url=url,
+            stream=True,
+            post=kwargs.get("post", False),
+            parameters=kwargs.get("parameters"),
+            timeout=kwargs.get("timeout"),
+            headers=headers,
+            encoding=kwargs.get("encoding"),
+        )
+        return self.stream_path(
+            path, "Download of %s failed in retrieval of stream!" % url
+        )
+
+
+class AzureBlobUpload:
+    """Wrapper for Azure Blob upload logic"""
+
+    def upload_file(
+        self,
+        dataset_name: str,
+        filename: str,
+        account: str,
+        container: str,
+        key: str,
+        data: None,
```

Comment:
It can be helpful to users of the function specify a type hint eg. Optional[Dict] if it can be None or a dictionary